### ISSUE TITLE
End2end Transformer Training by Using Ragged Tensor 

### ISSUE THREAD

**Issue by yaochengji** [tensorflow/tensorflow#40965]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (tf nightly):
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
I'd like to write a [transformer model](https://www.tensorflow.org/tutorials/text/transformer) by end2end using Ragged Tensor to speed up training. However, I found that there're some ops that Ragged Tensor could not support, ex. `matmul`, `relu`.  Do you have plan to support the other tf ops for Ragged Tensor?

In general, there are two kinds of ops to support:
1. Ragged Tensor op could reuse the kernel written for regular tensor. ex. `relu`
2. Ragged Tensor op could not reuse the kernel written for regular tensor. ex. `logits = tf.einsum("BTNH,BFNH->BNFT", key, query)`, where dimension `F` and `T` are ragged.

**Will this change the current api? How?** No.

**Who will benefit with this feature?**

Those models with batches of variable-length sequential inputs, ex. Transformer.

**Any Other info.**

**Comment by ARozental** [comment]
Additionally, there is a ragged op that is needed to do positional encoding and does't exist. 
Basically what I'm looking for is the option to chop a dense tensor to the shape of a ragged one and apply an op such as '+'.
Example:
```
text: "Google is nice"
embedded_text: [[1.1,2.2],[3.3,4.4],[5.5,6.6]] #ragged tensor after text embedding, <hidden_size=2, text_length=3>
positional_encoding_matrix: [[0.0,1.0],[0.1,0.2],[0.5,0.6],[0.8,0.2],[0.4,0.9]]] #<hidden_size=2, max_text_length=5>
wanted_function(embedded_text,positional_encoding_matrix,'+'): 
  step1: chops the second input to the shape of the ragged tensor, [[0.0,1.0],[0.1,0.2],[0.5,0.6],[0.8,0.2],[0.4,0.9]]]=>[[0.0,1.0],[0.1,0.2],[0.5,0.6]]
  step2: apply the '+' op on the equally shaped ragged tensors and return the result
```
This would enable us to use the Transformer model without adding padding vectors (and the attention masks that come with it).

I'm starting to work on a version of a ragged transformer now but I still don't see a way to write the above function efficiently without help from the TF team.

**Comment by byronyi** [comment]
https://github.com/tensorflow/community/pull/260 might be relevant.

**Comment by edloper** [comment]
@yaochengji For elementwise functions such as `relu`, you can just use `tf.ragged.map_flat_values`.  E.g.:

```
x = tf.ragged.constant([[-3.0, 2.0], [3.0]])
tf.ragged.map_flat_values(tf.nn.relu, x)
```

Adding ragged support to `tf.einsum` would be more complicated; any contributions are welcome. 

@ARozental You can chop a dense tensor to the shape of a ragged one with something like this:

```
def chop_dense_to_ragged(dense, ragged):
  mask = ragged.with_flat_values(tf.ones([tf.shape(ragged.flat_values)[0]], tf.bool))
  mask = mask.to_tensor(shape=tf.shape(dense)[:tf.rank(mask)])
  return tf.ragged.boolean_mask(dense, mask)

d = tf.reshape(tf.range(12), [3, 4])
r = tf.ragged.constant([[1, 2], [3], [4, 5, 6]])
print(chop_dense_to_ragged(d, r))
# prints <tf.RaggedTensor [[0, 1], [4], [8, 9, 10]]>
print(r + chop_dense_to_ragged(d, r))
# prints <tf.RaggedTensor [[1, 3], [7], [12, 14, 16]]>
```
